### PR TITLE
[Analytics Plugin] Export withAnalyticsContext

### PR DESCRIPTION
**Pull request by aitherios**
## Hey, I just made a Pull Request!

The `withAnalyticsContext` function exists in the `@backstage/core-plugin-api` package without being exported. This patch intends to put it to use and expose this alternative to use the analytics plugin.

The function had to me altered a bit and because of data it was covered with tests in the same fashion as the `AnalyticsContext` component.

### Design choice

The `withAnalyticsContext` function has a `TProps extends {} = any` to allow receiving components that the type signature cannot be inferred.

```js
export function withAnalyticsContext<TProps extends {} = any>(
  Component: React.ComponentType<TProps>,
  attributes: Partial<AnalyticsContextValue>,
) {
'...'
}
```
This would allow, for example:
```js
withAnalyticsContext(
  ({ prop }) => <div data-testid="prop">{prop}</div>,
  { extension: 'test' },
);
```

But if type analysis needs to be always enforced, the alternative with:
```js
export function withAnalyticsContext<TProps extends {}>(
  Component: React.ComponentType<TProps>,
  attributes: Partial<AnalyticsContextValue>,
) {
'...'
}
```
Would force the function user to write code like, for example:
```js
withAnalyticsContext(
  ({ prop }: { prop: string }) => <div data-testid="prop">{prop}</div>,
  { extension: 'test' },
);
```

The former was chosen just for the ease of usage. If the type error on the typescript compiler was clear I wouldn't see this as an issue but it is not as straight forward as it could be.

I'm not sure what is the Backstage code base policy in this kind of issue and I appreciate a nudge in the right direction =].

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by aitherios** [comment]
I'm running into this issue while running `yarn build:api-reports packages/core-plugin-api`

```bash
$ yarn build:api-reports packages/core-plugin-api
# Compiling TypeScript
# Generating package API reports
## Processing packages/core-plugin-api

Internal Error: Unable to analyze the export "withAnalyticsContext" in
/Users/renanm/Source/work/gh/backstage/dist-types/packages/core-plugin-api/src/analytics/AnalyticsContext.d.ts

You have encountered a software defect. Please consider reporting the issue to the maintainers of this application.
```

File: packages/core-plugin-api/src/analytics/AnalyticsContext.tsx [hunk 1]
```diff
@@ -85,19 +85,19 @@ export const AnalyticsContext = (options: {
...
+ * the given attributes.
  *
  * @param Component - Component to be wrapped with analytics context attributes
- * @param values - Analytics context key/value pairs.
- * @internal
+ * @param attributes - Analytics context key/value pairs.
+ * @public
  */
-export function withAnalyticsContext<TProps extends {}>(
+export function withAnalyticsContext<TProps extends {} = any>(
```
Comment (Rugvip) [inline packages/core-plugin-api/src/analytics/AnalyticsContext.tsx:94]: Don't think we want to default this to `any`, how come you changed it?

File: packages/core-plugin-api/src/analytics/AnalyticsContext.tsx [hunk 2]
```diff
@@ -85,19 +85,19 @@ export const AnalyticsContext = (options: {
...
   Component: React.ComponentType<TProps>,
-  values: AnalyticsContextValue,
+  attributes: Partial<AnalyticsContextValue>,
 ) {
```
Comment (Rugvip) [inline packages/core-plugin-api/src/analytics/AnalyticsContext.tsx:97]: Could be nice to add an explicit return type here to clean things up, although we don't really have a precedent for that

File: .changeset/nasty-ways-fail.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/core-plugin-api': patch
+---
+
+Export withAnalyticsContext high order component
```
Comment (Rugvip) [inline .changeset/nasty-ways-fail.md:5]:
```suggestion
Export `withAnalyticsContext` high order component
```

**Review by Rugvip — CHANGES_REQUESTED** [review]
Yep seems like this was the intention from the beginning, thanks! 👍

**Pull request closed without merging**